src/app/reporting/ReportScheduler.h

```diff
@@ -0,0 +1,150 @@
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/ReadHandler.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/IntrusiveList.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+// Forward declaration of TestReportScheduler to allow it to be friend with ReportScheduler
+class TestReportScheduler;
+
+using Timestamp = System::Clock::Timestamp;
+
+class ReportScheduler : public ReadHandler::Observer
+{
+public:
+    /// @brief This class acts as an interface between the report scheduler and the system timer to reduce dependencies on the
+    /// system layer.
+    class TimerDelegate
+    {
+    public:
+        virtual ~TimerDelegate() {}
+        /// @brief Start a timer for a given context, the report scheduler will always start by trying to cancel an existing timer
+        /// before starting a new one
+        /// @param context context to pass to the timer callback, in this case, a read handler node is passed
+        /// @param aTimeout time in miliseconds before the timer expires
+        virtual CHIP_ERROR StartTimer(void * context, System::Clock::Timeout aTimeout) = 0;
+        /// @brief Cancel a timer for a given context
+        /// @param context used to identify the timer to cancel
+        virtual void CancelTimer(void * context)         = 0;
+        virtual bool IsTimerActive(void * context)       = 0;
+        virtual Timestamp GetCurrentMonotonicTimestamp() = 0;
+    };
+
+    class ReadHandlerNode : public IntrusiveListNodeBase<>
+    {
+    public:
+        using TimerCompleteCallback = void (*)();
+
+        ReadHandlerNode(ReadHandler * aReadHandler, TimerDelegate * aTimerDelegate, TimerCompleteCallback aCallback) :
+            mTimerDelegate(aTimerDelegate), mCallback(aCallback)
+        {
+            VerifyOrDie(aReadHandler != nullptr);
+            VerifyOrDie(aTimerDelegate != nullptr);
+            VerifyOrDie(aCallback != nullptr);
+
+            mReadHandler = aReadHandler;
+            SetIntervalTimeStamps(aReadHandler);
+        }
+        ReadHandler * GetReadHandler() const { return mReadHandler; }
+        /// @brief Check if the Node is reportable now, meaning its readhandler was made reportable by attribute dirtying and
+        /// handler state, and minimal time interval last since last report has elapsed, or the maximal time interval since last
```

Comment:
```suggestion
        /// handler state, and minimal time interval since last report has elapsed, or the maximal time interval since last
```
Yes, I missed that last pass when trying to fix up the grammar here...